### PR TITLE
first milestone has been completed 

### PR DESCRIPTION
**Pull request by usman-ghani3**
Two ne api's created 

- data/semantic_search
- data/chatbot

### CODE REVIEW

File: intelliserver/api/functions/data/index.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+const keys = {
+    'openai': process.env.OPENAI_API_KEY,
+    'replicate': process.env.REPLICATE_API_KEY,
+    'sagemaker': process.env.SAGEMAKER_API_KEY,
+    'cohere': process.env.COHERE_API_KEY
+};
+
+async function callSemanticSearchAPI(data) {
+    try {
+        const apiUrl = 'https://1cw9da6wl7.execute-api.us-east-2.amazonaws.com/semantic_search/';
```
Comment (intelligentnode) [inline intelliserver/api/functions/data/index.js:18]: read the api from .env file with default url.

something like:
apiUrl  = os.environ.get('INTELLI_SEARCH_URL', 'https://1cw9da6wl7.execute-api.us-east-2.amazonaws.com/semantic_search/')

File: intelliserver/api/functions/data/index.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+const { USE_DEFAULT_KEYS } = require(path.join(global.__basedir, 'config'));
+
+const keys = {
+    'openai': process.env.OPENAI_API_KEY,
+    'replicate': process.env.REPLICATE_API_KEY,
+    'sagemaker': process.env.SAGEMAKER_API_KEY,
+    'cohere': process.env.COHERE_API_KEY
+};
+
+async function callSemanticSearchAPI(data) {
```
Comment (intelligentnode) [inline intelliserver/api/functions/data/index.js:16]: add swager documentation. you can use openai to auto generate the documentation.

check example under: function/chatbot/index.js

File: intelliserver/api/functions/data/index.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        const functions = req.body.functions;
+        const function_call = req.body.function_call;
+        const results = await chatbot.chat(input, functions, function_call);
+
+        res.json({ status: "OK", data: results });
+    } catch (error) {
+        res.json({ status: "ERROR", message: error.message });
+    }
+});
+
```
Comment (intelligentnode) [inline intelliserver/api/functions/data/index.js:94]: create another variation of the chatbot called stream. similar to functions/chatbot/index.js

Comment (intelligentnode) [reply intelliserver/api/functions/data/index.js:94]: router.post('/stream', async (req, res, next) => {
....

File: intelliserver/api/functions/data/index.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+router.post('/chatbot', async (req, res, next) => {
+    try {
+        // Call the semantic search API asynchronously
+        const semanticSearchResponse = await callSemanticSearchAPI(req.body);
+
+        // Concatenate the text from response data
+        const concatenatedText = semanticSearchResponse.data.map(document => {
+            return document.data.map(item => item.text).join(' ');
+        }).join(' ');
+        req.body.input.messages[0].content = concatenatedText;
```
Comment (intelligentnode) [inline intelliserver/api/functions/data/index.js:80]: you should send the user query and the response of the semantic search to the chatbot as follow:

context: <semantic_search_response>

user query: <user_query>


best to add text file template in public or root, then load it and replace the place holders.

**Review by intelligentnode — CHANGES_REQUESTED** [review]

File: intelliserver/api/functions/data/index.js (hunk 2 above)
Comment (usman-ghani3) [reply intelliserver/api/functions/data/index.js:18]: feedback implemented

File: intelliserver/api/functions/data/index.js (hunk 1 above)
Comment (usman-ghani3) [reply intelliserver/api/functions/data/index.js]: feedback implemented

File: intelliserver/api/functions/data/index.js (hunk 2 above)
Comment (usman-ghani3) [reply intelliserver/api/functions/data/index.js:18]: feedback implemented

File: intelliserver/api/functions/data/index.js (hunk 4 above)
Comment (usman-ghani3) [reply intelliserver/api/functions/data/index.js]: feedback implemented

File: intelliserver/api/functions/data/index.js (hunk 3 above)
Comment (usman-ghani3) [reply intelliserver/api/functions/data/index.js:77]: Stream api created and modified according to the requirements

**Pull request closed without merging**